modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modeling
+
+import (
+	"container/list"
+	"fmt"
+	"math"
+)
+
+// LocalSliceAlgorithm 'borrows' and 'rents' endpoints from other zones to make
+// the local EndpointSlice balanced with the incoming traffic (number of nodes
+// distribution)
+type LocalSliceAlgorithm struct{}
+
+type endpointDeviation struct {
+	name      string
+	deviation float64
+}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy
+func (alg LocalSliceAlgorithm) CreateSliceGroups(region regionInfo) (map[string]EndpointSliceGroup, error) {
+	if region.zoneDetails == nil {
+		return nil, fmt.Errorf("can't create EndpointSlices with 0 number of zone")
+	}
+	sliceGroups := map[string]EndpointSliceGroup{}
+	// stack stores 'endpointDeviation' -- the info of how should number of
+	// endpoints in a zone be adjusted to match the ratio balance
+	endpointStack := list.New()
+	for name, zone := range region.zoneDetails {
+		var localGroup EndpointSliceGroup
+		localGroup.Label = name
+		localGroup.Weights = map[string]float64{name: 1.0}
+		localGroup.Composition = map[string]int{}
+
+		deviation := math.Round(zone.nodesRatio*float64(region.totalEndpoints) - float64(zone.Endpoints))
+		// assign the endpoints to its local sliceGroup
+		localGroup.Composition[name] = int(math.Min(float64(zone.Endpoints), float64(zone.Endpoints)+deviation))
+		if deviation == 0 {
+			continue
+		}
+		sliceGroups[name] = localGroup
+		err := createGroupHelper(endpointStack, endpointDeviation{name: name, deviation: deviation}, sliceGroups)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// all endpoints should have been distributed properly
+	if endpointStack.Len() != 0 {
+		return nil, fmt.Errorf("stack of endpoints should be empty but with size %d", endpointStack.Len())
+	}
+	return sliceGroups, nil
+}
+
+// createGroupHelper helps to 'borrow' or 'rent' endpoints to other
+// EndpointSliceGroups
+func createGroupHelper(stack *list.List, receiver endpointDeviation, sliceGroups map[string]EndpointSliceGroup) error {
```

Comment:
I may be wrong here, but this seems like it could be simplified. I think all we actually need to do for this entire algorithm is calculate the expected endpoints for each zone given the proportion of nodes in the zone and calculating the same proportion of endpoints. Then you can simply loop through zones once to get all the endpoints available to give from each zone, and then distribute them to zones that need them in a second loop. Not sure if that makes sense here, but I think it could end up being simpler to implement and understand.

Example pseudocode:

```
endpointsAvailable := map[string]int{}
for zone in zones {
  zone.deviation := zone.endpoints - expectedEndpoints(zone)
  if zone.deviation > 0 {
    endpointsAvailable[zone.name] := zone.deviation
  }
}

for zone in zones {
  if zone.deviation < 0 {
    takeEndpoints(zone, endpointsAvailable)
  }
}
```

I know it's more complicated than that pseudocode, but maybe that general approach would be simpler?

Reply:
Agreed, this seems to be a neater implementation and much more straightforward compared to the previous one which is quite intricate to understand at the first glance.

Reply:
Real implementation looks a little bit different from the pseudo code above, but the idea remains the same